db/merge_helper.cc

```diff
@@ -57,120 +58,297 @@ MergeHelper::MergeHelper(Env* env, const Comparator* user_comparator,
+Status MergeHelper::TimedFullMergeImpl(
+    const MergeOperator* merge_operator, const Slice& key,
+    MergeOperator::MergeOperationInputV3::ExistingValue&& existing_value,
+    const std::vector<Slice>& operands, Logger* logger, Statistics* statistics,
+    SystemClock* clock, bool update_num_ops_stats, std::string* result,
+    Slice* result_operand, ValueType* result_type,
+    MergeOperator::OpFailureScope* op_failure_scope) {
+  assert(result);
+  assert(result_type);
+
+  auto visitor = overload{
+      [&](std::string&& new_value) -> Status {
+        *result_type = kTypeValue;
+
+        if (result_operand) {
+          *result_operand = Slice(nullptr, 0);
+        }
+
+        *result = std::move(new_value);
+
+        return Status::OK();
+      },
+      [&](MergeOperator::MergeOperationOutputV3::NewColumns&& new_columns)
+          -> Status {
+        *result_type = kTypeWideColumnEntity;
+
+        if (result_operand) {
+          *result_operand = Slice(nullptr, 0);
+        }
+
+        result->clear();
+
+        WideColumns sorted_columns;
+        sorted_columns.reserve(new_columns.size());
+
+        for (const auto& column : new_columns) {
+          sorted_columns.emplace_back(column.first, column.second);
+        }
+
+        WideColumnsHelper::SortColumns(sorted_columns);
+
+        return WideColumnSerialization::Serialize(sorted_columns, *result);
+      },
+      [&](Slice&& operand) {
+        *result_type = kTypeValue;
+
+        if (result_operand) {
+          *result_operand = operand;
+          result->clear();
+        } else {
+          result->assign(operand.data(), operand.size());
+        }
+
+        return Status::OK();
+      }};
+
+  return TimedFullMergeCommonImpl(merge_operator, key,
+                                  std::move(existing_value), operands, logger,
+                                  statistics, clock, update_num_ops_stats,
+                                  op_failure_scope, std::move(visitor));
+}
+
+Status MergeHelper::TimedFullMergeImpl(
+    const MergeOperator* merge_operator, const Slice& key,
+    MergeOperator::MergeOperationInputV3::ExistingValue&& existing_value,
+    const std::vector<Slice>& operands, Logger* logger, Statistics* statistics,
+    SystemClock* clock, bool update_num_ops_stats, std::string* result_value,
+    PinnableWideColumns* result_entity,
+    MergeOperator::OpFailureScope* op_failure_scope) {
+  assert(result_value || result_entity);
+  assert(!result_value || !result_entity);
+
+  auto visitor = overload{
+      [&](std::string&& new_value) -> Status {
+        if (result_value) {
+          *result_value = std::move(new_value);
+
+          return Status::OK();
+        }
+
+        assert(result_entity);
+        result_entity->SetPlainValue(std::move(new_value));
+
+        return Status::OK();
+      },
+      [&](MergeOperator::MergeOperationOutputV3::NewColumns&& new_columns)
+          -> Status {
+        if (result_value) {
```

Comment:
I'm trying to think of a good use case where merge is done with columns while the user wants to get the result back as value. And the other way around, too.

I'm wondering if we should encourage users to provide `result_value` only if the new value is plain OR `result_entity` only if the new value is columns.

Reply:
This is the `TimedFullMergeImpl` variant that is used in point lookups like `Get`/`MultiGet`/`GetEntity`/`MultiGetEntity`. The thing about these APIs is that the API itself determines the format of the output: callers of `Get` and `MultiGet` need a plain value, while callers of `GetEntity` and `MultiGetEntity` need a wide-column structure. For backward compatibility, as well as to prevent chicken-and-egg problems due to clients having no way of knowing whether a key is associated with a plain KV or an entity, we do format translation for them, e.g. for `Get` calls, we just return the value of the default column values of any entities, and for `GetEntity` calls, any plain values get wrapped in a single-column entity.

Note that the above behavior is not actually new; it's just that the implementation got migrated to `TimedFullMerge` from `MemTable` / `GetContext`.